models/User.go

```diff
@@ -1408,75 +1408,147 @@ func PerformMortgageTransaction(userId, stockId uint32, stockQuantity int64) (*T
+		err = db.Raw(sql, user.Id, stockId).Scan(&stockCount).Error
 		if err != nil {
+			l.Infof("PerformMortgageTransaction failed for userId = %d, stockId = %d amount = %d while retrieving total stocks", userId, stockId, stockQuantity)
 			l.Error(err)
 			return nil, err
 		}
 
-		// Sc will be negative if stocks are there in mortgage.
-		// Change sign to mean number of stocks in mortgage
-		stockCount.Sc *= -1
-
-		l.Infof("%d stocks mortgaged currently", stockCount.Sc)
+		l.Debugf("%d stocks mortgaged currently", stockCount.Sc)
 
 		if stockQuantity > stockCount.Sc {
 			l.Errorf("Insufficient stocks in mortgage. Have %d, want %d", stockCount.Sc, stockQuantity)
 			return nil, NotEnoughStocksError{}
 		}
 
 	} else {
-		MortgagePutLimitRWMutex.RLock()
-		lim := int64(MortgagePutLimit)
-		MortgagePutLimitRWMutex.RUnlock()
-		if user.Total > lim {
-			return nil, WayTooMuchCashError{}
-		}
-
-		rate = MORTGAGE_DEPOSIT_RATE
-		l.Debugf("stockQuantity negative. Depositing stocks to mortgage @ %d%%", rate)
+		l.Debugf("Mortgaging stocks in action")
 
+		// stockOwned = Total no. of stocks user owns
 		stockOwned, err := getSingleStockCount(user, stockId)
 		if err != nil {
 			l.Error(err)
 			return nil, err
 		}
 
-		if stockOwned < -stockQuantity {
-			l.Errorf("Insufficient stocks in ownership. Have %d, want %d", stockOwned, stockQuantity)
+		if utils.AbsInt64(stockQuantity) > stockOwned {
+			l.Errorf("Insufficient stocks to mortgage. Have %d, want %d", stockOwned, stockQuantity)
 			return nil, NotEnoughStocksError{}
 		}
 	}
 
-	trTotal := -int64(mortgagePrice) * stockQuantity * int64(rate) / 100
-	l.Infof("lalalalalalalala %d %d %d", trTotal, stockQuantity, int64(rate/100))
-	if int64(user.Cash)+trTotal < 0 {
-		l.Debugf("User does not have enough cash. Want %d, Have %d. Failing.", trTotal, user.Cash)
-		return nil, NotEnoughCashError{}
+	var trTotal int64
+
+	/* It is retrive action; This block will set correct trTotal and stockQuantity equal
+	to amount of stocks that could be retrieved; This block also modifies MortgageDetails
+	table, btw this table is not used anywhere except in current function */
+
+	if stockQuantity >= 0 {
+
+		db := getDB()
+		sql := "SELECT id, stocksInBank, mortgagePrice from MortgageDetails where userId=? AND stockId=? ORDER BY mortgagePrice"
+		rows, err := db.Raw(sql, user.Id, stockId).Rows()
+		if err != nil {
+			l.Error(err)
+			return nil, err
+		}
+		defer rows.Close()
+
+		// price represents cost of stock at time of mortgage
+		// stocksInBank represents number of stocks mortgaged at that time
+		var id int32
+		var price int64
+		var stocksInBank int64
+
+		// tempStockQuantity is stock quantity left to retrieve
+		tempStockQuantity := stockQuantity
+		tempUserCash := int64(user.Cash)
+		l.Infof("Cash %d", tempUserCash)
+		var maxStocksRetrieval int64
+
+		for rows.Next() {
+			rows.Scan(&id, &stocksInBank, &price)
+			maxStocksRetrieval = tempUserCash / price
```

Comment:
I feel this logic can be simplified a lot. Breaking out of loop etc feels like a complication that can be avoided. A simple way to have done this would have been

```python
while stocks_to_retrieve > 0:
    id, stocks_in_bank, price = retrieve_record_from_db()
    if stocks_in_bank < stocks_to_retrieve:
        ...
    elif stocks_in_bank > stocks_to_retrieve:
        ...
    else stocks_in_bank == stocks_to_retrieve:
        ...
```
I mean sure the method you've used might work too. But incase something goes wrong later, I feel this is easier to pinpoint.